File: src/types/itemize.ts
export type TransferType =
  | "Sent"
  | "Received"
  | "LocalChange"
  | "NoUpdate"
  | "Message";

export type FileType =
  | "File"
  | "Directory"
  | "Symlink"
  | "Device"
  | "Special";

export type DifferenceKind =
  | "Checksum"
  | "Size"
  | "Timestamp"
  | "Permissions"
  | "Owner"
  | "Group"
  | "Acl"
  | "ExtendedAttributes"
  | "NewlyCreated";

export interface ItemizedChange {
  transfer_type: TransferType;
  file_type: FileType;
  differences: DifferenceKind[];
  path: string;
}

File: crates/rsync-core/src/models/execution/itemize.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ItemizedChange {
    pub transfer_type: TransferType,
    pub file_type: FileType,
    pub differences: Vec<DifferenceKind>,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum TransferType {
    Sent,        // >
    Received,    // <
    LocalChange, // c
    NoUpdate,    // .
    Message,     // * (e.g., *deleting)
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum FileType {
    File,      // f
    Directory, // d
    Symlink,   // L
    Device,    // D
    Special,   // S
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum DifferenceKind {
    Checksum,           // c (position 2)
    Size,               // s (position 3)
    Timestamp,          // t (position 4)
    Permissions,        // p (position 5)
    Owner,              // o (position 6)
    Group,              // g (position 7)
    Acl,                // a (position 9)
    ExtendedAttributes, // x (position 10)
    NewlyCreated,       // + in any position
}
